vm/src/stdlib/itertools.rs

```diff
@@ -488,6 +490,196 @@ mod decl {
+                false
+            }
+        }
+    }
+
+    #[pyclass(name = "groupby")]
+    struct PyItertoolsGroupBy {
+        iterable: PyObjectRef,
+        key_func: Option<PyObjectRef>,
+        state: Mutex<GroupByState>,
+    }
+
+    impl PyValue for PyItertoolsGroupBy {
+        fn class(vm: &VirtualMachine) -> PyClassRef {
+            vm.class("itertools", "groupby")
+        }
+    }
+
+    impl fmt::Debug for PyItertoolsGroupBy {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("PyItertoolsGroupBy")
+                .field("iterable", &self.iterable)
+                .field("key_func", &self.key_func)
+                .field("state", &self.state.lock())
+                .finish()
+        }
+    }
+
+    #[derive(FromArgs)]
+    struct GroupByArgs {
+        iterable: PyObjectRef,
+        #[pyarg(positional_or_keyword, optional = true)]
+        key: OptionalOption<PyObjectRef>,
+    }
+
+    #[pyimpl]
+    impl PyItertoolsGroupBy {
+        #[pyslot]
+        fn tp_new(
+            cls: PyClassRef,
+            args: GroupByArgs,
+            vm: &VirtualMachine,
+        ) -> PyResult<PyRef<Self>> {
+            let iter = get_iter(vm, &args.iterable)?;
+
+            PyItertoolsGroupBy {
+                iterable: iter,
+                key_func: args.key.flat_option(),
+                state: Mutex::new(GroupByState {
+                    current_key: None,
+                    current_value: None,
+                    next_group: false,
+                    grouper: None,
+                }),
+            }
+            .into_ref_with_type(vm, cls)
+        }
+
+        #[pymethod(name = "__next__")]
+        fn next(
+            zelf: PyRef<Self>,
+            vm: &VirtualMachine,
+        ) -> PyResult<(PyObjectRef, PyItertoolsGrouperRef)> {
+            let mut state = zelf.state.lock();
+            state.grouper = None;
+
+            if !state.next_group {
+                let current_key = state.current_key.clone();
+                drop(state);
+
+                let (value, key) = if let Some(old_key) = current_key {
+                    loop {
+                        let (value, new_key) = zelf.advance(vm)?;
+                        if !vm.bool_eq(new_key.clone(), old_key.clone())? {
+                            break (value, new_key);
+                        }
+                    }
+                } else {
+                    zelf.advance(vm)?
+                };
+
+                state = zelf.state.lock();
+                state.current_value = Some(value);
+                state.current_key = Some(key);
+            }
+
+            state.next_group = false;
+
+            let grouper = PyItertoolsGrouper {
+                groupby: zelf.clone(),
+            }
+            .into_ref(vm);
+
+            state.grouper = Some(Arc::downgrade(&grouper.clone().into_typed_pyobj()));
+            Ok((state.current_key.clone().unwrap(), grouper))
```

Comment:
This unwrapping implies curernt_key is always `Some`. In that case, I suggest to use `unwrap().clone()` rather than `clone().unwrap()` because the intention is cloning the key object rather than the `Option` itself.